analyzers/src/SonarAnalyzer.CSharp/Rules/ReferenceEqualityCheckWhenEqualsExists.cs

```diff
@@ -80,105 +77,60 @@ protected override void Initialize(SonarAnalysisContext context)
 
                             var typeLeft = c.SemanticModel.GetTypeInfo(binary.Left).Type;
                             var typeRight = c.SemanticModel.GetTypeInfo(binary.Right).Type;
-                            if (typeLeft == null ||
-                                typeRight == null ||
-                                IsAllowedType(typeLeft) ||
-                                IsAllowedType(typeRight))
+                            if (IsAllowedTypeOrNull(typeLeft) || IsAllowedTypeOrNull(typeRight))
                             {
                                 return;
                             }
 
                             if (MightOverrideEquals(typeLeft, allInterfacesWithImplementationsOverriddenEquals) ||
                                 MightOverrideEquals(typeRight, allInterfacesWithImplementationsOverriddenEquals))
                             {
-                                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, binary.OperatorToken.GetLocation()));
+                                c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, binary.OperatorToken.GetLocation()));
                             }
                         },
                         SyntaxKind.EqualsExpression,
                         SyntaxKind.NotEqualsExpression);
                 });
-        }
 
-        private static bool MightOverrideEquals(ITypeSymbol type, ISet<INamedTypeSymbol> allInterfacesWithImplementationsOverriddenEquals)
-        {
-            return HasEqualsOverride(type) ||
-                allInterfacesWithImplementationsOverriddenEquals.Contains(type) ||
-                HasTypeConstraintsWhichMightOverrideEquals(type, allInterfacesWithImplementationsOverriddenEquals);
-        }
+        private static bool MightOverrideEquals(ITypeSymbol type, ISet<INamedTypeSymbol> allInterfacesWithImplementationsOverriddenEquals) =>
+            HasEqualsOverride(type)
+            || allInterfacesWithImplementationsOverriddenEquals.Contains(type)
+            || HasTypeConstraintsWhichMightOverrideEquals((ITypeParameterSymbol)type, allInterfacesWithImplementationsOverriddenEquals);
 
-        private static bool HasTypeConstraintsWhichMightOverrideEquals(ITypeSymbol type, ISet<INamedTypeSymbol> allInterfacesWithImplementationsOverriddenEquals)
-        {
-            if (type.TypeKind != TypeKind.TypeParameter)
-            {
-                return false;
-            }
+        private static bool HasTypeConstraintsWhichMightOverrideEquals(ITypeParameterSymbol type, ISet<INamedTypeSymbol> allInterfacesWithImplementationsOverriddenEquals) =>
+            type.ConstraintTypes.Any(x => MightOverrideEquals(x, allInterfacesWithImplementationsOverriddenEquals));
```

Comment:
We need the `type is ITypeParameterSymbol typeParameter` check. Nancy IT fails with:
>CSC : error AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.ReferenceEqualityCheckWhenEqualsExists' threw an exception of type 'System.InvalidCastException' with message 'Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Symbols.PublicModel.NonErrorNamedTypeSymbol' to type 'Microsoft.CodeAnalysis.ITypeParameterSymbol'.'. [C:\ProgramData\vsts-agent\1\s\analyzers\its\sources\Nancy\src\Nancy\Nancy.csproj]

Please add this test case:
```
interface IWithGeneric
{
    void Something<T>();
}

class UsingInterfaces
{
    public void Compare(IWithGeneric a, IWithGeneric b)
    {
        if (a == b) { }
    }
}
```

Reply:
@pavel-mikula-sonarsource Thanks! I say that one, but my googling did not find the right solution. Added the case, and the check as well.